server/src/test/java/org/elasticsearch/cluster/NodeConnectionsServiceTests.java

```diff
@@ -106,46 +105,55 @@ public void testConnectAndDisconnect() throws Exception {
+                    // if the previous iteration was a disrupting one then there could still be some pending disconnections which would
+                    // prevent us from asserting that all nodes are connected in this iteration without this call.
+                    ensureConnections(service);
                 }
-            });
-            disruptionThread.start();
-
-            final DiscoveryNodes nodes = discoveryNodesFromList(randomSubsetOf(allNodes));
-            final PlainActionFuture<Void> future = new PlainActionFuture<>();
-            service.connectToNodes(nodes, () -> future.onResponse(null));
-            future.actionGet();
-            if (isDisrupting == false) {
-                assertConnected(nodes);
-            }
-            service.disconnectFromNodesExcept(nodes);
+                final AtomicBoolean stopDisrupting = new AtomicBoolean();
+                final Thread disruptionThread = new Thread(() -> {
+                    while (isDisrupting && stopDisrupting.get() == false) {
+                        transportService.disconnectFromNode(randomFrom(allNodes));
+                    }
+                }, "disruption thread " + iteration);
+                disruptionThread.start();
+
+                final DiscoveryNodes nodes = discoveryNodesFromList(randomSubsetOf(allNodes));
+                final PlainActionFuture<Void> future = new PlainActionFuture<>();
+                service.connectToNodes(nodes, () -> future.onResponse(null));
+                future.actionGet();
+                if (isDisrupting == false) {
+                    assertConnected(nodes);
+                }
+                service.disconnectFromNodesExcept(nodes);
 
-            assertTrue(stopDisrupting.compareAndSet(false, true));
-            disruptionThread.join();
+                assertTrue(stopDisrupting.compareAndSet(false, true));
+                disruptionThread.join();
 
-            if (randomBoolean()) {
-                // sometimes do not wait for the disconnections to complete before starting the next connections
-                if (usually()) {
-                    ensureConnections(service);
-                    assertConnectedExactlyToNodes(nodes);
-                } else {
-                    assertBusy(() -> assertConnectedExactlyToNodes(nodes));
+                if (randomBoolean()) {
+                    // sometimes do not wait for the disconnections to complete before starting the next connections
+                    if (usually()) {
+                        ensureConnections(service);
+                        assertConnectedExactlyToNodes(nodes);
+                    } else {
+                        assertBusy(() -> assertConnectedExactlyToNodes(nodes));
+                    }
                 }
             }
+        } finally {
+            assertTrue(stopReconnecting.compareAndSet(false, true));
+            reconnectionThread.join();
+            ensureConnections(service);
```

Comment:
I would prefer to put ensureConnections outside the finally block to ensure that if the test fails, we see the original error rather than potentially an error from ensureConnections.